File: go/test/employee_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/kota-sdk/go"
	"github.com/voxgig-sdk/kota-sdk/go/core"

	vs "github.com/voxgig-sdk/kota-sdk/go/utility/struct"
)

func TestEmployeeEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Employee(nil)
		if ent == nil {
			t.Fatal("expected non-nil EmployeeEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"employee": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Employee(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Employee(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := employeeBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "employee." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set KOTA_TEST_EMPLOYEE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		employeeRef01Ent := client.Employee(nil)
		employeeRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "employee"}, setup.data), "employee_ref01"))

		employeeRef01DataResult, err := employeeRef01Ent.Create(employeeRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		employeeRef01Data = core.ToMapAny(entityData(employeeRef01DataResult))
		if employeeRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if employeeRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		employeeRef01Match := map[string]any{}

		employeeRef01ListResult, err := employeeRef01Ent.List(employeeRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		employeeRef01List, employeeRef01ListOk := employeeRef01ListResult.([]any)
		if !employeeRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", employeeRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(employeeRef01List), map[string]any{"id": employeeRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		employeeRef01DataUp0Up := map[string]any{
			"id": employeeRef01Data["id"],
		}

		employeeRef01MarkdefUp0Name := "date_of_birth"
		employeeRef01MarkdefUp0Value := fmt.Sprintf("Mark01-employee_ref01_%d", setup.now)
		employeeRef01DataUp0Up[employeeRef01MarkdefUp0Name] = employeeRef01MarkdefUp0Value

		employeeRef01ResdataUp0Result, err := employeeRef01Ent.Update(employeeRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		employeeRef01ResdataUp0 := core.ToMapAny(entityData(employeeRef01ResdataUp0Result))
		if employeeRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if employeeRef01ResdataUp0["id"] != employeeRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if employeeRef01ResdataUp0[employeeRef01MarkdefUp0Name] != employeeRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", employeeRef01MarkdefUp0Name, employeeRef01ResdataUp0[employeeRef01MarkdefUp0Name])
		}

		// LOAD
		employeeRef01MatchDt0 := map[string]any{
			"id": employeeRef01Data["id"],
		}
		employeeRef01DataDt0Loaded, err := employeeRef01Ent.Load(employeeRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		employeeRef01DataDt0LoadResult := core.ToMapAny(entityData(employeeRef01DataDt0Loaded))
		if employeeRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if employeeRef01DataDt0LoadResult["id"] != employeeRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func employeeBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "employee", "EmployeeTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read employee test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse employee test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"employee01", "employee02", "employee03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("KOTA_TEST_EMPLOYEE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"KOTA_TEST_EMPLOYEE_ENTID": idmap,
		"KOTA_TEST_LIVE":      "FALSE",
		"KOTA_TEST_EXPLAIN":   "FALSE",
		"KOTA_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["KOTA_TEST_EMPLOYEE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["KOTA_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["KOTA_APIKEY"],
			},
			extra,
		})
		client = sdk.NewKotaSDK(core.ToMapAny(mergedOpts))
	}

	live := env["KOTA_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["KOTA_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/EmployeeEntityTest.php
<?php
declare(strict_types=1);

// Employee entity test

require_once __DIR__ . '/../kota_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class EmployeeEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = KotaSDK::test(null, null);
        $ent = $testsdk->Employee(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "employee" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = KotaSDK::test($seed, null);
        $seen = iterator_to_array($base->Employee(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = KotaConfig::make_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = KotaSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->Employee(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = employee_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "update", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "employee." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set KOTA_TEST_EMPLOYEE_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $employee_ref01_ent = $client->Employee(null);
        $employee_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.employee"), "employee_ref01"));

        $employee_ref01_data_result = $employee_ref01_ent->create($employee_ref01_data, null);
        $employee_ref01_data = Helpers::to_map(is_object($employee_ref01_data_result) && method_exists($employee_ref01_data_result, 'data_get') ? $employee_ref01_data_result->data_get() : $employee_ref01_data_result);
        $this->assertNotNull($employee_ref01_data);
        $this->assertNotNull($employee_ref01_data["id"]);

        // LIST
        $employee_ref01_match = [];

        $employee_ref01_list_result = $employee_ref01_ent->list($employee_ref01_match, null);
        $this->assertIsArray($employee_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($employee_ref01_list_result),
            ["id" => $employee_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // UPDATE
        $employee_ref01_data_up0_up = [
            "id" => $employee_ref01_data["id"],
        ];

        $employee_ref01_markdef_up0_name = "date_of_birth";
        $employee_ref01_markdef_up0_value = "Mark01-employee_ref01_" . $setup["now"];
        $employee_ref01_data_up0_up[$employee_ref01_markdef_up0_name] = $employee_ref01_markdef_up0_value;

        $employee_ref01_resdata_up0_result = $employee_ref01_ent->update($employee_ref01_data_up0_up, null);
        $employee_ref01_resdata_up0 = Helpers::to_map(is_object($employee_ref01_resdata_up0_result) && method_exists($employee_ref01_resdata_up0_result, 'data_get') ? $employee_ref01_resdata_up0_result->data_get() : $employee_ref01_resdata_up0_result);
        $this->assertNotNull($employee_ref01_resdata_up0);
        $this->assertEquals($employee_ref01_resdata_up0["id"], $employee_ref01_data_up0_up["id"]);
        $this->assertEquals($employee_ref01_resdata_up0[$employee_ref01_markdef_up0_name], $employee_ref01_markdef_up0_value);

        // LOAD
        $employee_ref01_match_dt0 = [
            "id" => $employee_ref01_data["id"],
        ];
        $employee_ref01_data_dt0_loaded = $employee_ref01_ent->load($employee_ref01_match_dt0, null);
        $employee_ref01_data_dt0_load_result = Helpers::to_map(is_object($employee_ref01_data_dt0_loaded) && method_exists($employee_ref01_data_dt0_loaded, 'data_get') ? $employee_ref01_data_dt0_loaded->data_get() : $employee_ref01_data_dt0_loaded);
        $this->assertNotNull($employee_ref01_data_dt0_load_result);
        $this->assertEquals($employee_ref01_data_dt0_load_result["id"], $employee_ref01_data["id"]);

    }
}

function employee_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/employee/EmployeeTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = KotaSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["employee01", "employee02", "employee03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("KOTA_TEST_EMPLOYEE_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "KOTA_TEST_EMPLOYEE_ENTID" => $idmap,
        "KOTA_TEST_LIVE" => "FALSE",
        "KOTA_TEST_EXPLAIN" => "FALSE",
        "KOTA_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["KOTA_TEST_EMPLOYEE_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["KOTA_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["KOTA_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new KotaSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["KOTA_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["KOTA_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
